From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bastiaan van der Plaat <bastiaan.v.d.plaat@gmail.com>
Date: Mon, 12 Oct 2026 18:09:00 +0200
Subject: [PATCH] Speed up the collision

The ARM9 has no floating point unit, the collision compared floats
through the software floating point functions. The comparisons of the
traces, the leaves of a box and the edicts in an area are integer
comparisons of the bits of the floats, in the order of their values,
which are exactly the same without NaNs. The corner of a box against a
brush side is picked by the signs of its normal, the box hull of the
entities has their signs too. These functions run from ITCM as ARM code.
---
 game/q_shared.c   | 10 ++++----
 game/q_shared.h   | 16 ++++++++++--
 qcommon/cmodel.c  | 62 +++++++++++++++++++++++------------------------
 server/sv_world.c | 18 +++++++-------
 4 files changed, 58 insertions(+), 48 deletions(-)

diff --git a/game/q_shared.c b/game/q_shared.c
index 90b376f..bebe0e3 100644
--- a/game/q_shared.c
+++ b/game/q_shared.c
@@ -346,7 +346,7 @@ Returns 1, 2, or 1 + 2
 ==================
 */
 #if !id386 || defined __linux__ 
-int BoxOnPlaneSide (vec3_t emins, vec3_t emaxs, struct cplane_s *p)
+HOT_CODE int BoxOnPlaneSide (vec3_t emins, vec3_t emaxs, struct cplane_s *p)
 {
 	float	dist1, dist2;
 	int		sides;
@@ -354,9 +354,9 @@ int BoxOnPlaneSide (vec3_t emins, vec3_t emaxs, struct cplane_s *p)
 // fast axial cases
 	if (p->type < 3)
 	{
-		if (p->dist <= emins[p->type])
+		if (F_LE (p->dist, emins[p->type]))
 			return 1;
-		if (p->dist >= emaxs[p->type])
+		if (F_GE (p->dist, emaxs[p->type]))
 			return 2;
 		return 3;
 	}
@@ -403,9 +403,9 @@ dist2 = p->normal[0]*emaxs[0] + p->normal[1]*emaxs[1] + p->normal[2]*emaxs[2];
 	}
 
 	sides = 0;
-	if (dist1 >= p->dist)
+	if (F_GE (dist1, p->dist))
 		sides = 1;
-	if (dist2 < p->dist)
+	if (F_LT (dist2, p->dist))
 		sides |= 2;
 
 	assert( sides != 0 );
diff --git a/game/q_shared.h b/game/q_shared.h
index a662178..164de97 100644
--- a/game/q_shared.h
+++ b/game/q_shared.h
@@ -152,6 +152,18 @@ extern long Q_ftol( float f );
 #endif
 
 #define DotProduct(x,y)			(x[0]*y[0]+x[1]*y[1]+x[2]*y[2])
+
+// Hot code goes to ITCM, the fast memory of the DS for code, as ARM code
+#define	HOT_CODE			__attribute__((section(".itcm"), target("arm"), noinline))
+
+// Comparisons of floats that aren't NaN as integers, without the software
+// floating point functions: the bits of a float in the order of its value,
+// with -0 like 0. A macro, ARM code can't inline a Thumb function.
+#define	Q_FLOAT_ORDER(x)	({ union { float f; int i; } u_ = { (x) }; u_.i < 0 ? -(u_.i & 0x7fffffff) : u_.i; })
+#define	F_GT(a,b)	(Q_FLOAT_ORDER (a) > Q_FLOAT_ORDER (b))
+#define	F_GE(a,b)	(Q_FLOAT_ORDER (a) >= Q_FLOAT_ORDER (b))
+#define	F_LT(a,b)	(Q_FLOAT_ORDER (a) < Q_FLOAT_ORDER (b))
+#define	F_LE(a,b)	(Q_FLOAT_ORDER (a) <= Q_FLOAT_ORDER (b))
 #define VectorSubtract(a,b,c)	(c[0]=a[0]-b[0],c[1]=a[1]-b[1],c[2]=a[2]-b[2])
 #define VectorAdd(a,b,c)		(c[0]=a[0]+b[0],c[1]=a[1]+b[1],c[2]=a[2]+b[2])
 #define VectorCopy(a,b)			(b[0]=a[0],b[1]=a[1],b[2]=a[2])
@@ -189,11 +201,11 @@ float LerpAngle (float a1, float a2, float frac);
 #define BOX_ON_PLANE_SIDE(emins, emaxs, p)	\
 	(((p)->type < 3)?						\
 	(										\
-		((p)->dist <= (emins)[(p)->type])?	\
+		F_LE ((p)->dist, (emins)[(p)->type])?	\
 			1								\
 		:									\
 		(									\
-			((p)->dist >= (emaxs)[(p)->type])?\
+			F_GE ((p)->dist, (emaxs)[(p)->type])?\
 				2							\
 			:								\
 				3							\
diff --git a/qcommon/cmodel.c b/qcommon/cmodel.c
index b28bc68..fdf7677 100644
--- a/qcommon/cmodel.c
+++ b/qcommon/cmodel.c
@@ -770,7 +770,7 @@ void CM_InitBoxHull (void)
 
 		p = &box_planes[i*2+1];
 		p->type = 3 + (i>>1);
-		p->signbits = 0;
+		p->signbits = 1 << (i>>1);	// the clipping picks the corner of the box by it
 		VectorClear (p->normal);
 		p->normal[i>>1] = -1;
 	}	
@@ -810,7 +810,7 @@ CM_PointLeafnum_r
 
 ==================
 */
-int CM_PointLeafnum_r (vec3_t p, int num)
+HOT_CODE int CM_PointLeafnum_r (vec3_t p, int num)
 {
 	float		d;
 	cnode_t		*node;
@@ -825,7 +825,7 @@ int CM_PointLeafnum_r (vec3_t p, int num)
 			d = p[plane->type] - plane->dist;
 		else
 			d = DotProduct (plane->normal, p) - plane->dist;
-		if (d < 0)
+		if (F_LT (d, 0))
 			num = node->children[1];
 		else
 			num = node->children[0];
@@ -857,7 +857,7 @@ int		*leaf_list;
 float	*leaf_mins, *leaf_maxs;
 int		leaf_topnode;
 
-void CM_BoxLeafnums_r (int nodenum)
+HOT_CODE void CM_BoxLeafnums_r (int nodenum)
 {
 	cplane_t	*plane;
 	cnode_t		*node;
@@ -999,7 +999,7 @@ qboolean	trace_ispoint;		// optimized case
 CM_ClipBoxToBrush
 ================
 */
-void CM_ClipBoxToBrush (vec3_t mins, vec3_t maxs, vec3_t p1, vec3_t p2,
+HOT_CODE void CM_ClipBoxToBrush (vec3_t mins, vec3_t maxs, vec3_t p1, vec3_t p2,
 					  trace_t *trace, cbrush_t *brush)
 {
 	int			i, j;
@@ -1037,10 +1037,9 @@ void CM_ClipBoxToBrush (vec3_t mins, vec3_t maxs, vec3_t p1, vec3_t p2,
 
 			// push the plane out apropriately for mins/maxs
 
-			// FIXME: use signbits into 8 way lookup for each mins/maxs
 			for (j=0 ; j<3 ; j++)
 			{
-				if (plane->normal[j] < 0)
+				if (plane->signbits & (1 << j))
 					ofs[j] = maxs[j];
 				else
 					ofs[j] = mins[j];
@@ -1056,23 +1055,23 @@ void CM_ClipBoxToBrush (vec3_t mins, vec3_t maxs, vec3_t p1, vec3_t p2,
 		d1 = DotProduct (p1, plane->normal) - dist;
 		d2 = DotProduct (p2, plane->normal) - dist;
 
-		if (d2 > 0)
+		if (F_GT (d2, 0))
 			getout = true;	// endpoint is not in solid
-		if (d1 > 0)
+		if (F_GT (d1, 0))
 			startout = true;
 
 		// if completely in front of face, no intersection
-		if (d1 > 0 && d2 >= d1)
+		if (F_GT (d1, 0) && F_GE (d2, d1))
 			return;
 
-		if (d1 <= 0 && d2 <= 0)
+		if (F_LE (d1, 0) && F_LE (d2, 0))
 			continue;
 
 		// crosses face
-		if (d1 > d2)
+		if (F_GT (d1, d2))
 		{	// enter
 			f = (d1-DIST_EPSILON) / (d1-d2);
-			if (f > enterfrac)
+			if (F_GT (f, enterfrac))
 			{
 				enterfrac = f;
 				clipplane = plane;
@@ -1082,7 +1081,7 @@ void CM_ClipBoxToBrush (vec3_t mins, vec3_t maxs, vec3_t p1, vec3_t p2,
 		else
 		{	// leave
 			f = (d1+DIST_EPSILON) / (d1-d2);
-			if (f < leavefrac)
+			if (F_LT (f, leavefrac))
 				leavefrac = f;
 		}
 	}
@@ -1094,11 +1093,11 @@ void CM_ClipBoxToBrush (vec3_t mins, vec3_t maxs, vec3_t p1, vec3_t p2,
 			trace->allsolid = true;
 		return;
 	}
-	if (enterfrac < leavefrac)
+	if (F_LT (enterfrac, leavefrac))
 	{
-		if (enterfrac > -1 && enterfrac < trace->fraction)
+		if (F_GT (enterfrac, -1) && F_LT (enterfrac, trace->fraction))
 		{
-			if (enterfrac < 0)
+			if (F_LT (enterfrac, 0))
 				enterfrac = 0;
 			trace->fraction = enterfrac;
 			trace->plane = *clipplane;
@@ -1113,7 +1112,7 @@ void CM_ClipBoxToBrush (vec3_t mins, vec3_t maxs, vec3_t p1, vec3_t p2,
 CM_TestBoxInBrush
 ================
 */
-void CM_TestBoxInBrush (vec3_t mins, vec3_t maxs, vec3_t p1,
+HOT_CODE void CM_TestBoxInBrush (vec3_t mins, vec3_t maxs, vec3_t p1,
 					  trace_t *trace, cbrush_t *brush)
 {
 	int			i, j;
@@ -1137,10 +1136,9 @@ void CM_TestBoxInBrush (vec3_t mins, vec3_t maxs, vec3_t p1,
 
 		// push the plane out apropriately for mins/maxs
 
-		// FIXME: use signbits into 8 way lookup for each mins/maxs
 		for (j=0 ; j<3 ; j++)
 		{
-			if (plane->normal[j] < 0)
+			if (plane->signbits & (1 << j))
 				ofs[j] = maxs[j];
 			else
 				ofs[j] = mins[j];
@@ -1151,7 +1149,7 @@ void CM_TestBoxInBrush (vec3_t mins, vec3_t maxs, vec3_t p1,
 		d1 = DotProduct (p1, plane->normal) - dist;
 
 		// if completely in front of face, no intersection
-		if (d1 > 0)
+		if (F_GT (d1, 0))
 			return;
 
 	}
@@ -1168,7 +1166,7 @@ void CM_TestBoxInBrush (vec3_t mins, vec3_t maxs, vec3_t p1,
 CM_TraceToLeaf
 ================
 */
-void CM_TraceToLeaf (int leafnum)
+HOT_CODE void CM_TraceToLeaf (int leafnum)
 {
 	int			k;
 	int			brushnum;
@@ -1237,7 +1235,7 @@ CM_RecursiveHullCheck
 
 ==================
 */
-void CM_RecursiveHullCheck (int num, float p1f, float p2f, vec3_t p1, vec3_t p2)
+HOT_CODE void CM_RecursiveHullCheck (int num, float p1f, float p2f, vec3_t p1, vec3_t p2)
 {
 	cnode_t		*node;
 	cplane_t	*plane;
@@ -1249,7 +1247,7 @@ void CM_RecursiveHullCheck (int num, float p1f, float p2f, vec3_t p1, vec3_t p2)
 	int			side;
 	float		midf;
 
-	if (trace_trace.fraction <= p1f)
+	if (F_LE (trace_trace.fraction, p1f))
 		return;		// already hit something nearer
 
 	// if < 0, we are in a leaf node
@@ -1292,26 +1290,26 @@ return;
 #endif
 
 	// see which sides we need to consider
-	if (t1 >= offset && t2 >= offset)
+	if (F_GE (t1, offset) && F_GE (t2, offset))
 	{
 		CM_RecursiveHullCheck (node->children[0], p1f, p2f, p1, p2);
 		return;
 	}
-	if (t1 < -offset && t2 < -offset)
+	if (F_LT (t1, -offset) && F_LT (t2, -offset))
 	{
 		CM_RecursiveHullCheck (node->children[1], p1f, p2f, p1, p2);
 		return;
 	}
 
 	// put the crosspoint DIST_EPSILON pixels on the near side
-	if (t1 < t2)
+	if (F_LT (t1, t2))
 	{
 		idist = 1.0/(t1-t2);
 		side = 1;
 		frac2 = (t1 + offset + DIST_EPSILON)*idist;
 		frac = (t1 - offset + DIST_EPSILON)*idist;
 	}
-	else if (t1 > t2)
+	else if (F_GT (t1, t2))
 	{
 		idist = 1.0/(t1-t2);
 		side = 0;
@@ -1326,9 +1324,9 @@ return;
 	}
 
 	// move up to the node
-	if (frac < 0)
+	if (F_LT (frac, 0))
 		frac = 0;
-	if (frac > 1)
+	if (F_GT (frac, 1))
 		frac = 1;
 		
 	midf = p1f + (p2f - p1f)*frac;
@@ -1339,9 +1337,9 @@ return;
 
 
 	// go past the node
-	if (frac2 < 0)
+	if (F_LT (frac2, 0))
 		frac2 = 0;
-	if (frac2 > 1)
+	if (F_GT (frac2, 1))
 		frac2 = 1;
 		
 	midf = p1f + (p2f - p1f)*frac2;
diff --git a/server/sv_world.c b/server/sv_world.c
index 6d9cd1e..43a1a08 100644
--- a/server/sv_world.c
+++ b/server/sv_world.c
@@ -351,7 +351,7 @@ SV_AreaEdicts_r
 
 ====================
 */
-void SV_AreaEdicts_r (areanode_t *node)
+HOT_CODE void SV_AreaEdicts_r (areanode_t *node)
 {
 	link_t		*l, *next, *start;
 	edict_t		*check;
@@ -372,12 +372,12 @@ void SV_AreaEdicts_r (areanode_t *node)
 
 		if (check->solid == SOLID_NOT)
 			continue;		// deactivated
-		if (check->absmin[0] > area_maxs[0]
-		|| check->absmin[1] > area_maxs[1]
-		|| check->absmin[2] > area_maxs[2]
-		|| check->absmax[0] < area_mins[0]
-		|| check->absmax[1] < area_mins[1]
-		|| check->absmax[2] < area_mins[2])
+		if (F_GT (check->absmin[0], area_maxs[0])
+		|| F_GT (check->absmin[1], area_maxs[1])
+		|| F_GT (check->absmin[2], area_maxs[2])
+		|| F_LT (check->absmax[0], area_mins[0])
+		|| F_LT (check->absmax[1], area_mins[1])
+		|| F_LT (check->absmax[2], area_mins[2]))
 			continue;		// not touching
 
 		if (area_count == area_maxcount)
@@ -394,9 +394,9 @@ void SV_AreaEdicts_r (areanode_t *node)
 		return;		// terminal node
 
 	// recurse down both sides
-	if ( area_maxs[node->axis] > node->dist )
+	if ( F_GT (area_maxs[node->axis], node->dist) )
 		SV_AreaEdicts_r ( node->children[0] );
-	if ( area_mins[node->axis] < node->dist )
+	if ( F_LT (area_mins[node->axis], node->dist) )
 		SV_AreaEdicts_r ( node->children[1] );
 }
 
